lib/vector-core/src/event/log_event.rs

```diff
@@ -19,13 +23,40 @@ use super::{
 };
 use crate::{config::log_schema, event::MaybeAsLogMut, ByteSizeOf};
 
-#[derive(Clone, Debug, PartialEq, PartialOrd, Deserialize)]
+#[derive(Debug, Deserialize)]
 pub struct LogEvent {
     #[serde(flatten)]
     fields: Arc<Value>,
 
     #[serde(skip)]
     metadata: EventMetadata,
+
+    #[serde(skip)]
+    byte_size_cache: Atomic<Option<NonZeroUsize>>,
```

Comment:
Does the type logic here require this to be an atomic? If this were an `Option<usize>` we might squeeze some more juice out here.

Reply:
I had first written it as `Cell<Option<NonZeroUsize>>`, but this struct must be `Sync` due to trait bounds on some of the protobuf stuff IIRC, so that drives the `Atomic` implementation here.
```
error[E0277]: `Cell<std::option::Option<NonZeroUsize>>` cannot be shared between threads safely
   --> lib/vector-core/src/fanout.rs:13:23
    |
13  |     Add(ComponentKey, BufferSender<EventArray>),
    |                       ^^^^^^^^^^^^^^^^^^^^^^^^ `Cell<std::option::Option<NonZeroUsize>>` cannot be shared between threads safely
    |
    = help: within `LogEvent`, the trait `Sync` is not implemented for `Cell<std::option::Option<NonZeroUsize>>`
note: required because it appears within the type `LogEvent`
   --> lib/vector-core/src/event/log_event.rs:26:12
    |
26  | pub struct LogEvent {
    |            ^^^^^^^^
    = note: required because of the requirements on the impl of `Sync` for `Unique<LogEvent>`
    = note: required because it appears within the type `prost::alloc::raw_vec::RawVec<LogEvent>`
    = note: required because it appears within the type `Vec<LogEvent>`
note: required because it appears within the type `event::array::EventArray`
   --> lib/vector-core/src/event/array.rs:127:10
    |
127 | pub enum EventArray {
    |          ^^^^^^^^^^
    = note: required because of the requirements on the impl of `Bufferable` for `event::array::EventArray`
note: required by a bound in `BufferSender`
   --> /home/arch/vector/lib/vector-buffers/src/topology/channel/sender.rs:152:28
    |
152 | pub struct BufferSender<T: Bufferable> {
    |                            ^^^^^^^^^^ required by this bound in `BufferSender`
```

Reply:
Ah ha. Figured, but worth checking. Thanks for confirming.